core/src/main/scala/org/apache/spark/executor/Executor.scala

```diff
@@ -227,6 +228,17 @@ private[spark] class Executor(
               logError(errMsg)
             }
           }
+
+          if (releasedLocks.nonEmpty) {
+            val errMsg =
+              s"${releasedLocks.size} block locks were not released by TID = $taskId:\n" +
```

Comment:
In my production environment,  when the storage memory is full,  there is a great probability of deadlock.  It is a temporary patch because JoshRosen add a read/write lock for block in https://github.com/apache/spark/pull/10705 for Spark 2.0. 

Two theads are removing the same block which result in deadlock.
BlockManager will first lock MemoryManager and wait to lock BlockInfo in function 'dropFromMemory', Execturo task lock BlockInfo and wait to lock MemoryManager calling 'memstore.remove(block)' in function 'removeBlock' or function 'removeOldBlocks'.

So just a ConcurrentHashMap to record the locks by tasks. In case of failure, release all lock after task complete.